File: mcp-code-intelligence-nodejs/src/analyzers/complexity/counters/TypeScriptCounter.ts
/**
 * KSA-161: TypeScript/JavaScript decision point counter.
 */

import type { SyntaxNode } from '../../../parsers/types.js';
import { BaseNodeCounter } from './BaseNodeCounter.js';

export class TypeScriptCounter extends BaseNodeCounter {
  readonly language = 'typescript';
  readonly branchNodeTypes = [
    'if_statement',
    'switch_case',
    'ternary_expression',
  ];
  readonly loopNodeTypes = [
    'for_statement',
    'for_in_statement',
    'while_statement',
    'do_statement',
  ];
  readonly logicalOperators = ['&&', '||', '??'];
  readonly exceptionNodeTypes = ['catch_clause'];

  protected isLogicalOp(node: SyntaxNode): boolean {
    // TypeScript uses binary_expression for && and ||
    if (node.type === 'binary_expression') {
      const op = node.childForFieldName('operator');
      if (op && this.logicalOperators.includes(op.text)) return true;
    }
    // Optional chaining counted as a branch decision
    if (node.type === 'optional_chain_expression') return true;
    return false;
  }
}

File: backend/mcp-code-intelligence-kotlin/bin/main/com/codeintel/analyzers/complexity/counters/TypeScriptCounter.kt
/**
 * KSA-161: TypeScript/JavaScript AST node counter.
 */
package com.codeintel.analyzers.complexity.counters

import com.codeintel.analyzers.complexity.BaseNodeCounter

class TypeScriptCounter : BaseNodeCounter() {
    override val language = "typescript"
    override val branchNodeTypes = setOf(
        "if_statement", "switch_case", "ternary_expression",
        "conditional_expression",
    )
    override val loopNodeTypes = setOf(
        "for_statement", "for_in_statement", "while_statement",
        "do_statement",
    )
    override val logicalOperators = setOf("&&", "||", "??")
    override val exceptionNodeTypes = setOf("catch_clause")
}
